File: src/lib/backend/tauri/usage.ts
import type { UsageCollectResult, UsagePort } from '@/lib/backend/contracts';
import {
  mapCoreParserHealth,
  mapCoreUsageRecord,
  type CoreParserHealth,
  type CoreUsageRecord,
} from '@/lib/backend/contracts/usage-map';
import type {
  GatewayUsageOverview,
  GatewayUsageQuery,
  GatewayUsageRow,
  UsageAvailability,
  UsageOverview,
} from '@/lib/backend/contracts/usage-types';
import type { ParserHealth, UsageRecord, UsageTrendPoint } from '@/lib/types';
import { invoke } from './invoke';

/**
 * Production Usage — wired to core UsageService via Tauri commands.
 */
export function createTauriUsagePort(): UsagePort {
  return {
    async getAvailability(): Promise<UsageAvailability> {
      const v = await invoke<{ status: string; reason?: string }>('usage_get_availability');
      if (v.status === 'available') return { status: 'available' };
      return {
        status: 'unavailable',
        reason: v.reason ?? 'Usage backend 不可用',
      };
    },

    async queryUsage(q): Promise<UsageRecord[]> {
      const agentId = !q.agentId || q.agentId === 'all' ? null : q.agentId;
      const model = !q.model || q.model === 'all' ? null : q.model;
      const rows = await invoke<CoreUsageRecord[]>('usage_query', {
        days: q.days,
        agentId,
        model,
        limit: q.limit ?? null,
        since: q.since ?? null,
        excludeAgentIds: q.excludeAgentIds ?? null,
      });
      return rows.map(mapCoreUsageRecord);
    },

    async usageOverview(q): Promise<UsageOverview> {
      const agentId = !q.agentId || q.agentId === 'all' ? null : q.agentId;
      const model = !q.model || q.model === 'all' ? null : q.model;
      return invoke<UsageOverview>('usage_overview', {
        days: q.days,
        agentId,
        model,
        since: q.since ?? null,
        excludeAgentIds: q.excludeAgentIds ?? null,
      });
    },

    async usageTrend(days, agentId, model, since, excludeAgentIds, groupBy): Promise<UsageTrendPoint[]> {
      const id = !agentId || agentId === 'all' ? null : agentId;
      const modelFilter = !model || model === 'all' ? null : model;
      return invoke<UsageTrendPoint[]>('usage_trend', {
        days,
        agentId: id,
        model: modelFilter,
        since: since ?? null,
        excludeAgentIds: excludeAgentIds ?? null,
        groupBy: groupBy === 'model' ? 'model' : null,
      });
    },

    async listModels(): Promise<string[]> {
      return invoke<string[]>('usage_list_models');
    },

    async parserHealth(): Promise<ParserHealth[]> {
      const rows = await invoke<CoreParserHealth[]>('usage_parser_health');
      return rows.map(mapCoreParserHealth);
    },

    async missingPricingModels(days = 30): Promise<string[]> {
      return invoke<string[]>('usage_missing_pricing', { days });
    },

    async collectUsage(onProgress): Promise<UsageCollectResult> {
      onProgress?.(10);
      const result = await invoke<UsageCollectResult>('usage_collect', { agentId: null });
      onProgress?.(100);
      return result;
    },

    async gatewayUsageQuery(args = {}): Promise<GatewayUsageRow[]> {
      return gatewayUsageQuery(args);
    },

    async gatewayUsageOverview(args = {}): Promise<GatewayUsageOverview> {
      return gatewayUsageOverview(args);
    },
  };
}

export type {
  GatewayUsageOverview,
  GatewayUsageQuery,
  GatewayUsageRow,
} from '@/lib/backend/contracts/usage-types';

/**
 * Per-request usage observed by the local gateway (bridge), stored in its own
 * table (separate from the agent-log-derived usage records).
 */
export async function gatewayUsageQuery(
  args: GatewayUsageQuery = {},
): Promise<GatewayUsageRow[]> {
  return invoke<GatewayUsageRow[]>('gateway_usage_query', {
    since: args.since ?? null,
    until: args.until ?? null,
    profileId: args.profileId ?? null,
    limit: args.limit ?? null,
  });
}

/** Aggregated local gateway usage overview for a time window. */
export async function gatewayUsageOverview(
  args: GatewayUsageQuery = {},
): Promise<GatewayUsageOverview> {
  return invoke<GatewayUsageOverview>('gateway_usage_overview', {
    since: args.since ?? null,
    until: args.until ?? null,
    profileId: args.profileId ?? null,
  });
}

File: src-tauri/src/commands/usage.rs
//! Usage Tauri commands — collect / query / trend / models / health.

use agenthub_core::models::{
    AgentId, CollectResult, GatewayUsageOverview, GatewayUsageQuery, GatewayUsageRow, ParserHealth,
    UsageOverview, UsageQuery, UsageRecord,
};
use serde_json::Value;
use tauri::State;

use crate::commands::{map_err_string, parse_agent_opt, with_hub_blocking};
use crate::state::AppState;

#[tauri::command]
pub async fn usage_get_availability(state: State<'_, AppState>) -> Result<Value, String> {
    let _hub = state.hub_arc()?;
    // Wired in core; always available in Tauri desktop.
    Ok(serde_json::json!({ "status": "available" }))
}

#[tauri::command]
pub async fn usage_collect(
    state: State<'_, AppState>,
    agent_id: Option<String>,
) -> Result<CollectResult, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        let filter = parse_agent_opt(agent_id.as_deref())?;
        hub.usage()
            .collect(filter)
            .map_err(|e| map_err_string("usage_collect", e))
    })
    .await
}

#[tauri::command]
pub async fn usage_query(
    state: State<'_, AppState>,
    days: u32,
    agent_id: Option<String>,
    model: Option<String>,
    limit: Option<u32>,
    since: Option<String>,
    exclude_agent_ids: Option<Vec<String>>,
    until: Option<String>,
) -> Result<Vec<UsageRecord>, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        let agent = parse_agent_opt(agent_id.as_deref())?;
        let model = model.filter(|m| !m.is_empty() && m != "all");
        let since = since.filter(|s| !s.is_empty());
        let until = until.filter(|s| !s.is_empty());
        let exclude_agent_ids = parse_exclude_agent_ids(exclude_agent_ids);
        hub.usage()
            .query(UsageQuery {
                days: days.max(1),
                agent_id: agent,
                model,
                limit,
                since,
                until,
                exclude_agent_ids,
            })
            .map_err(|e| map_err_string("usage_query", e))
    })
    .await
}

#[tauri::command]
pub async fn usage_trend(
    state: State<'_, AppState>,
    days: u32,
    agent_id: Option<String>,
    model: Option<String>,
    since: Option<String>,
    exclude_agent_ids: Option<Vec<String>>,
    group_by: Option<String>,
    until: Option<String>,
) -> Result<Vec<Value>, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        let agent = parse_agent_opt(agent_id.as_deref())?;
        let model = model.filter(|m| !m.is_empty() && m != "all");
        let since = since.filter(|s| !s.is_empty());
        let until = until.filter(|s| !s.is_empty());
        let exclude = parse_exclude_agent_ids(exclude_agent_ids);
        let by_model = group_by.as_deref() == Some("model");
        let points = if by_model {
            hub.usage().trend_by_model(
                days.max(1),
                agent,
                model.as_deref(),
                since.as_deref(),
                &exclude,
                until.as_deref(),
            )
        } else {
            hub.usage().trend(
                days.max(1),
                agent,
                model.as_deref(),
                since.as_deref(),
                &exclude,
                until.as_deref(),
            )
        }
        .map_err(|e| map_err_string("usage_trend", e))?;
        Ok(points.into_iter().map(|p| Value::Object(p.0)).collect())
    })
    .await
}

#[tauri::command]
pub async fn usage_overview(
    state: State<'_, AppState>,
    days: u32,
    agent_id: Option<String>,
    model: Option<String>,
    since: Option<String>,
    exclude_agent_ids: Option<Vec<String>>,
    until: Option<String>,
) -> Result<UsageOverview, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        let agent = parse_agent_opt(agent_id.as_deref())?;
        let model = model.filter(|m| !m.is_empty() && m != "all");
        let since = since.filter(|s| !s.is_empty());
        let until = until.filter(|s| !s.is_empty());
        let exclude = parse_exclude_agent_ids(exclude_agent_ids);
        hub.usage()
            .overview(
                days.max(1),
                agent,
                model.as_deref(),
                since.as_deref(),
                &exclude,
                until.as_deref(),
            )
            .map_err(|e| map_err_string("usage_overview", e))
    })
    .await
}

fn parse_exclude_agent_ids(ids: Option<Vec<String>>) -> Vec<AgentId> {
    ids.unwrap_or_default()
        .into_iter()
        .filter_map(|s| AgentId::parse(&s))
        .collect()
}

#[tauri::command]
pub async fn usage_list_models(state: State<'_, AppState>) -> Result<Vec<String>, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        hub.usage()
            .list_models()
            .map_err(|e| map_err_string("usage_list_models", e))
    })
    .await
}

#[tauri::command]
pub async fn usage_parser_health(state: State<'_, AppState>) -> Result<Vec<ParserHealth>, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        hub.usage()
            .parser_health()
            .map_err(|e| map_err_string("usage_parser_health", e))
    })
    .await
}

/// Models in recent usage_records lacking embedded pricing.
#[tauri::command]
pub async fn usage_missing_pricing(
    state: State<'_, AppState>,
    days: Option<u32>,
) -> Result<Vec<String>, String> {
    let hub = state.hub_arc()?;
    let d = days.unwrap_or(30);
    with_hub_blocking(hub, move |hub| {
        hub.usage()
            .missing_pricing_models(d)
            .map_err(|e| map_err_string("usage_missing_pricing", e))
    })
    .await
}

/// Per-request usage observed by the local bridge (separate `gateway_usage`
/// table; never merged into `usage_records`).
#[tauri::command]
pub async fn gateway_usage_query(
    state: State<'_, AppState>,
    since: Option<String>,
    until: Option<String>,
    profile_id: Option<String>,
    limit: Option<u32>,
) -> Result<Vec<GatewayUsageRow>, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        hub.usage()
            .gateway_usage_query(GatewayUsageQuery {
                since: since.filter(|s| !s.is_empty()),
                until: until.filter(|s| !s.is_empty()),
                profile_id: profile_id.filter(|p| !p.is_empty()),
                limit,
            })
            .map_err(|e| map_err_string("gateway_usage_query", e))
    })
    .await
}

/// Aggregated overview over the `gateway_usage` table for a time window.
#[tauri::command]
pub async fn gateway_usage_overview(
    state: State<'_, AppState>,
    since: Option<String>,
    until: Option<String>,
    profile_id: Option<String>,
) -> Result<GatewayUsageOverview, String> {
    let hub = state.hub_arc()?;
    with_hub_blocking(hub, move |hub| {
        hub.usage()
            .gateway_usage_overview(GatewayUsageQuery {
                since: since.filter(|s| !s.is_empty()),
                until: until.filter(|s| !s.is_empty()),
                profile_id: profile_id.filter(|p| !p.is_empty()),
                limit: None,
            })
            .map_err(|e| map_err_string("gateway_usage_overview", e))
    })
    .await
}
